File: praca-inzynierska-backend/application/src/main/kotlin/jwer/pracainzynierskabackend/model/dto/StatisticsEmployeeYearDto.kt
package jwer.pracainzynierskabackend.model.dto

import jwer.pracainzynierskabackend.model.entity.Shift

data class StatisticsEmployeeYearDto(
        val employee: EmployeeDto,
        val totalMonths: List<StatisticsDto> =
                Array(12) {StatisticsDto(it)}.toList(),
        val positions: List<StatisticsPositionYearDto> =
                employee.positions.map { StatisticsPositionYearDto(it) },
        val total: StatisticsDto = StatisticsDto(null)
) {

    fun add(shift: Shift) {
        val month = shift.period.start.monthValue - 1
        totalMonths[month] += shift
        positions.find { it.position.id == shift.position.id }?.let {it.add(shift)}
        total += shift
    }

}

File: praca-inzynierska-frontend/src/app/app/model/dto/StatisticsEmployeeYearDto.ts
import { StatisticsDto } from './StatisticsDto';
import { EmployeeDto } from './EmployeeDto';
import { StatisticsPositionYearDto } from './StatisticsPositionYearDto';
import { PositionDto } from './PositionDto';

export class StatisticsEmployeeYearDto {

    constructor(
        public employee: EmployeeDto,
        public totalMonths: StatisticsDto[],
        public positions: StatisticsPositionYearDto[],
        public total: StatisticsDto
    ) {}

    sort() {
        this.positions.sort((a, b) => PositionDto.compare(a.position, b.position))
    }

    static copyOf(other: StatisticsEmployeeYearDto): StatisticsEmployeeYearDto {
        return new StatisticsEmployeeYearDto(
            other.employee, 
            other.totalMonths.map(m => StatisticsDto.copyOf(m)),
            other.positions.map(p => StatisticsPositionYearDto.copyOf(p)),
            StatisticsDto.copyOf(other.total)
        )
    }
}